### ISSUE TITLE
Use semantic error values

### ISSUE THREAD

**Issue by Xiretza** [etesync/etebase-rs#27]
Currently, the [`Error` enum](https://github.com/etesync/etebase-rs/blob/90cbd67b96d39b6d91368e78416c4e7191691cd9/src/error.rs#L12-L46) is little more than a string along with a broad error category. This makes it effectively impossible to analyze and react to these errors programmatically, as this would require parsing error messages that are subject to change at any time. Additionally, a lot of information is simply discarded this way (depending on how the string is constructed).

A better solution would be to store semantic types in the `Error` variants, especially if the error originates in another library that provides its own error types. The error values can still easily be converted to strings using `Display`/`ToString` when they need to be shown to the user, but they can also be interpreted by code. [`thiserror`](https://crates.io/crates/thiserror) is helpful for concisely implementing all the related traits.

**Comment by tasn** [comment]
Yeah, I'm open to having better errors. It was just a quick way to not have to worry about how to structure the errors.
